### DISCUSSION TITLE
How could I use NextJS Image component when rendering an Image?

### DISCUSSION THREAD

**Discussion by philmetzger** [ueberdosis/tiptap#3445] [General Questions] [answered]
How could I use the nextjs Image component instead of a simple img tag for the Image extension?
I tried to create a CustomImage to extend Image. Then how could I render a component out?

e.g. instead of this:
```
renderHTML({ HTMLAttributes }) {
    return ['img', mergeAttributes(this.options.HTMLAttributes, HTMLAttributes)];
  },
```

I would like to do this:
```
renderHTML({ HTMLAttributes }) {
    return <Image
        src={url}
        alt="preview"
        width={420}
        height={420}
        priority={true}
    />
  },
```
Where Image is `import NextImage from 'next/image';`

**Comment by amirhhashemi** [comment, marked as answer]
You can achieve that with a NodeView. I haven't tested this, but at least it gives you a rough idea of how to implement it:
```tsx
import { Image as ImageExtension } from '@tiptap/extension-image'
import { ReactNodeViewRenderer } from '@tiptap/react'
import Image from 'next/image'

const CustomImage = ImageExtension.extend({
    addNodeView() {
        return ReactNodeViewRenderer(ImageWrapper)
    },
})

const ImageWrapper = (props) => {
  return (
    <NodeViewWrapper>
      <Image src={props.node.attrs.src} />
    </NodeViewWrapper>
  )
}
```
Now `next/image` is used to render an image node in the editor but still the editor HTML output (`editor.getHTML()`) is a simple <img> tag. If you want the HTML output to have all `next/image` attributes (loading="lazy" for example) you should implement a custom `renderHTML` method. Something like this (again, I haven't tested it):
```tsx
const CustomImage = ImageExtension.extend({
    renderHTML({ HTMLAttributes }) {
      return [
        "img",
        mergeAttributes(this.options.HTMLAttributes, { "loading": "lazy" }),
      ];
    },

    addNodeView() {
        return ReactNodeViewRenderer(ImageWrapper)
    },
})
```
This adds `loading="lazy"` to the HTML output. You might need something more complex.

Related docs:
https://tiptap.dev/guide/node-views/react
https://tiptap.dev/guide/custom-extensions/#render-html

**Reply by philmetzger** [reply]
Hey @ahhshm
This is something i tried however `addNodeView` does not seem to be called.
I added a `console.log` in `renderHTML` and `addNodeView` and I only see the `console.log` from `renderHTML`.

I am using it like this to get the html:
```
generateHTML(JSON.parse(text), [
        CustomImage,
        .....
```

**Reply by philmetzger** [reply]
Figured it out, got to use the `useEditor` hook!

**Comment by cosbgn** [comment]
Hi, how would you then call this from your template for example? I'm trying to do something very similar (a raw html block). My code looks like this:

```
import Code from '@tiptap/extension-code'
const Html = Code.extend({
	name:'Html',
	renderHTML ({ node }) {
		const div = document.createElement('div')
		div.dataset.type = 'raw'
		div.innerHTML = node.attrs.html
		return div
	}
})
```
However when I call `editor.chain().focus().toggleHtml().run()` I get that `toggleHtml is not a function` - What am I missing?

**Reply by amirhhashemi** [reply]
I'm not sure if I understand what you want to achieve here. I think you need a custom [Command](https://tiptap.dev/api/commands#chain-commands). `toggleHtml` command doesn't exist by default. you have to add it yourself.

**Reply by cosbgn** [reply]
So what I'm trying to do is to have a `rawHtml` block. Ideally inside the editor it would load as a code block, so it's easy to edit it (i.e. it doesn't render as html, it renders text inside a <code> tag. 

Then, when I call `editor.getHTML()`. It would actually render the html, pretty much replace `<code>` with `<div>`. 

This way I can have a raw HTML block for things like SVGs, Icons, embeds, particular tables, or anything else really. 

What I tried so far is a custom extension which extends the codeBlock. The extension would load the text inside a `<code>` block, then when I use `editor.getHTML()` I can use a regex to simply replace `<code> and </code> with <div> and </div>.` I think this is the easies way to accomplish what I want. My custom code block looks like this:

```js
import Code from '@tiptap/extension-code'
const Html = Code.extend({
	name:'Html',
	renderHTML ({ node }) {
		const code_div = document.createElement('code')
		code_div.dataset.type = 'raw'
		code_div.innerHTML = node.attrs.html
		return code_div
	}
})
```

Now what I can't figure out is how do I activate/deactivate this block? With the normal code block, I can use `editor.chain().focus().toggleCodeBlock().run()` which switches from a simple `<p>` to a `<code>` - Where do I register a function which does the same for this custom block?

**Comment by cosbgn** [comment]
Ok, so to make things easier, with my current setup, if I simply use the existing `codeBlock` and then when I want to reneder the html I simply do `html = html.replaceAll("<pre><code>", "<div>").replaceAll("</pre></code>", "</div>")` everything works perfectly, the only small issue I have the that the HTML is escaped so it doesn't get rendered. For example a SVG becomes:

```js
  <div>&lt;svg class="w-6 h-6" fill="none" stroke="currentColor" viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"&gt;&lt;path d="M12 14l9-5-9-5-9 5 9 5z"&gt;&lt;/path&gt;&lt;path d="M12 14l6.16-3.422a12.083 12.083 0 01.665 6.479A11.952 11.952 0 0012 20.055a11.952 11.952 0 00-6.824-2.998 12.078 12.078 0 01.665-6.479L12 14z"&gt;&lt;/path&gt;&lt;path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M12 14l9-5-9-5-9 5 9 5zm0 0l6.16-3.422a12.083 12.083 0 01.665 6.479A11.952 11.952 0 0012 20.055a11.952 11.952 0 00-6.824-2.998 12.078 12.078 0 01.665-6.479L12 14zm-4 6v-7.5l4-2.222"&gt;&lt;/path&gt;&lt;/svg&gt;</div>

```

So the only thing I really need is how to disable escaping so that it gets rendered as an actual SVG. Is this possible?
